src/engine/Server.cpp

```diff
@@ -492,6 +494,139 @@ Awaitable<void> Server::process(
       parsedHttpRequest.operation_);
 }
 
+// _____________________________________________________________________________
+Awaitable<void> Server::executeUpdate(
+    const ParsedQuery& query, const QueryExecutionTree& qet,
+    const ad_utility::Timer& requestTimer,
+    SharedCancellationHandle cancellationHandle) {
+  AD_CONTRACT_CHECK(query.hasUpdateClause());
+  auto updateClause = query.updateClause();
+  if (!std::holds_alternative<updateClause::GraphUpdate>(updateClause.op_)) {
+    throw std::runtime_error(
+        "Only INSERT/DELETE update operations are currently supported.");
+  }
+  auto graphUpdate = std::get<updateClause::GraphUpdate>(updateClause.op_);
+  auto res = qet.getResult(true);
+
+  auto& vocab = index_.getVocab();
+  // DeltaTriples transfers the IDs that are actually inserted to its own
+  // LocalVocab. This LocalVocab only contains IDs that are related to the
+  // template. Most of the IDs will be added to the DeltaTriples' LocalVocab. An
+  // ID will not be added if it belongs to a Quad with a variable that has no
+  // solutions.
+  LocalVocab localVocab = {};
+  // The GraphUpdate consist of a template of `SparqlTripleSimpleWithGraph` and
+  // in some cases also an GraphPattern. The template contains
+  // `TripleComponent`s (s, p and o) and a `Graph` (graph). First transform
+  // everything except Variables in the template to IDs.
+  using IdOrVariable = std::variant<Id, Variable>;
+  auto transformSparqlTripleComponent =
+      [&vocab, &localVocab](TripleComponent component) -> IdOrVariable {
+    if (component.isVariable()) {
+      return std::move(component.getVariable());
+    } else {
+      return std::move(component).toValueId(vocab, localVocab);
+    }
+  };
+  auto transformGraph =
+      [&vocab,
+       &localVocab](SparqlTripleSimpleWithGraph::Graph graph) -> IdOrVariable {
+    return std::visit(
+        ad_utility::OverloadCallOperator{
+            [](const std::monostate&) -> IdOrVariable {
+              AD_CORRECTNESS_CHECK(
+                  qlever::specialIds().contains(DEFAULT_GRAPH_IRI));
+              return qlever::specialIds().at(DEFAULT_GRAPH_IRI);
+            },
```

Comment:
That is not correct, these special IDs are only valid during the very first phase of index building.
You can get a normal IRI for the default graph from the normal vocab (e.g. transform it to a TripleComponent that stroes and IRI and then call toValueId etc.